storage/storage.go

```diff
@@ -230,30 +230,75 @@ type TupleBackend interface {
 	// ListObjectsByType returns all the objects of a specific type.
 	// You can assume that the type has already been validated.
 	// The result can't have duplicate elements.
-	ListObjectsByType(ctx context.Context, store string, objectType string) (ObjectIterator, error)
+	ListObjectsByType(
+		ctx context.Context,
+		store string,
+		objectType string,
+	) (ObjectIterator, error)
 
 	// ReadPage is similar to Read, but with PaginationOptions. Instead of returning a TupleIterator, ReadPage
 	// returns a page of tuples and a possibly non-empty continuation token.
-	ReadPage(context.Context, string, *openfgapb.TupleKey, PaginationOptions) ([]*openfgapb.Tuple, []byte, error)
+	ReadPage(
+		ctx context.Context,
+		store string,
+		tk *openfgapb.TupleKey,
+		opts PaginationOptions,
+	) ([]*openfgapb.Tuple, []byte, error)
 
 	// Write updates data in the tuple backend, performing all delete operations in
 	// `deletes` before adding new values in `writes`, returning the time of the transaction, or an error.
 	// It is expected that
 	// - there is at most 10 deletes/writes
 	// - no duplicate item in delete/write list
-	Write(context.Context, string, Deletes, Writes) error
-
-	ReadUserTuple(context.Context, string, *openfgapb.TupleKey) (*openfgapb.Tuple, error)
-
-	ReadUsersetTuples(context.Context, string, *openfgapb.TupleKey) (TupleIterator, error)
+	Write(ctx context.Context, store string, d Deletes, w Writes) error
+
+	ReadUserTuple(
+		ctx context.Context,
+		store string,
+		tk *openfgapb.TupleKey,
+	) (*openfgapb.Tuple, error)
+
+	ReadUsersetTuples(
+		ctx context.Context,
+		store string,
+		tk *openfgapb.TupleKey,
+	) (TupleIterator, error)
+
+	// ReverseReadTuples performs a reverse read of relationship tuples starting at one or
+	// more user(s) or userset(s) and filtered by object type and relation.
```

Comment:
The definition of Read API:

> // Read the set of tuples associated with `store` and `key`, which may be partially filled. A key must specify at
>	// least one of `Object` or `User` (or both), and may also optionally constrain by relation.

They both look very similar. 

Why is this one called "reverse"?

Reply:
As far as calling it "reverse" I don't think we need to keep that nomenclature if we don't want. But the query is really only intended for query relationship tuples starting at the user/userset and working backwards. It's meant to be indexed as such.

My understanding is that `Read` reads either all tuples for a single object (optionally constrained by a relation) or it reads all objects of any type for a user (optionally constrained by relation).

The goal here is to read all objects that a user or userset has a specific relation with without filtering by object_id whatsoever. As a simple query example,

`SELECT object_type, object_id FROM tuples WHERE store='mystore AND object_type='document' AND relation='viewer' AND user_type=X AND user=Y`